src/Native/LibTorchSharp/THSTensorFactories.cpp

```diff
@@ -170,43 +177,100 @@ Tensor THSTensor_new(
     void (*deleter)(void*),
     const int64_t* sizes,
     const int szlength,
-    int8_t scalar_type,
+    int8_t scalar_type,         // The element type in the data array
+    int8_t dtype,               // The element type of the constructed tensor
+    const int device_type,
+    const int device_index,
     const bool requires_grad)
 {
+    bool move = device_type != 0;
+    bool convert = scalar_type != dtype;
+
     auto options = at::TensorOptions()
         .dtype(at::ScalarType(scalar_type))
-        .requires_grad(requires_grad);
-
-    CATCH_TENSOR(torch::from_blob(data, at::ArrayRef<int64_t>(sizes, szlength), deleter, options));
+        .requires_grad(requires_grad && !move);
+
+    CATCH(
+        auto res = torch::from_blob(data, at::ArrayRef<int64_t>(sizes, szlength), deleter, options);
+        if (move) // Not CPU
+        {
+            auto device = c10::Device((c10::DeviceType)device_type, (c10::DeviceIndex)device_index);
+            res = convert
+                ? res.to(device, at::ScalarType(dtype), false, false).set_requires_grad(requires_grad)
+                : res.to(device, false, false).set_requires_grad(requires_grad);
```

Comment:
Just a style question. We prefer this over e.g.
```suggestion
            res = convert
                ? res.to(device, at::ScalarType(dtype), false, false)
                : res.to(device, false, false);
            res = res.set_requires_grad(requires_grad);
```

Reply:
Well, to a certain point (line length being reasonable), I personally like it, feels more functional, but I won't pretend that I'm consistent about it.